### ISSUE TITLE
Unable to use the SQL Server Maximum DateTime

### ISSUE THREAD

**Issue by MarkSonghurst** [minus5/gofreetds#18]
Hello.
My SQL Server database is refusing to accept the SQL Maximum Datetime of "9999-12-31 23:59:59.997" as a stored procedure parameter with the following error:

```
Msg 20018, Level 16\nGeneral SQL Server error: Check messages from the SQL Server\n\n\nMsg 8023, Level 16, State 1\nServer 'MUSTANG', Line 1\n\tThe incoming tabular data stream (TDS) remote procedure call (RPC) protocol stream is incorrect. Parameter 16 (\"@dtOwnershipEndDate\"): The supplied value is not a valid instance of data type datetime. Check the source data for invalid values. An example of an invalid value is data of numeric type with scale greater than precision.\n"
```

I am assigning the "9999-12-31 23:59:59.997" value to a Go time.Time (which appears ok), and FreeTDS/gofreetds is correctly recognising that the stored procedure parameter should be a SYBDATETIME.

Some logging I added to convert_sql_buf.go within function typeToSqlBuf showed that the Go Time value is broken down into these components:
```
diff -2643127431936722624 days -30591 secs 4275457715
```

I'll try and do so more research to help assist with this, but if you can see any immediate issues with using the SQL Maximum datetime from within gofreetds, I would be very grateful if you could let me know and I can work around it by using a lower datetime.

I'm using SQL Server 2014 and FreeTDS 0.95

Thanks.

**Comment by MarkSonghurst** [comment]
I think the UnixNano for the 9999-12-31 datetime is too big for an int64, so it's going negative which results in a negative days value (as shown in my comment above), which I suspect is the cause of the issue.

I'm struggling with the math in the SYBDATETIME case statement, but I think we need to use uint64 when dealing with extreme datetime values.
I wrote this function to get the uint64 of a time.Time

```
func UnixNano(t time.Time) uint64 {
	secs := t.Unix()
	var nanos uint64  
	nanos = (uint64(secs) * 1e9) + uint64(t.Nanosecond())
	return nanos
}
```

My intention was then to add a uint64 version of the current int64 code in the SYBDATETIME case, something like this:

```
if tm.String() == "9999-12-31 23:59:59.997 +0000 UTC" {
  sqlMaxTime := time.Date(9999, 12, 31, 23, 59, 59, 997, time.UTC)
  diff := UnixNano(sqlMaxTime) - UnixNano(sqlStartTime) // 15803616640775829989
  _, of := tm.Zone()
  diff += uint64(time.Duration(of) * time.Second)
  days := int32(diff / 1e9 / 60 / 60 / 24)
  secs := uint32(float64(diff-uint64(days)*1e9*60*60*24) * 0.0000003)
  err = binary.Write(buf, binary.LittleEndian, days)
  if err == nil {
    err = binary.Write(buf, binary.LittleEndian, secs)
  }
} else {
  //existing code which uses int64
}
```

I'm afraid my math is terrible and I'm struggling to get the intended result from this - within SQL Server the max datetime is being stored as "2400-10-18 05:30:40.773"

Any assistance would be appreciated. Thanks.

**Comment by MarkSonghurst** [comment]
Broken unit test now resolved in #24
